### DISCUSSION TITLE
TypeError: [E930] when trying to train textcat

### DISCUSSION THREAD

**Discussion by kevinlu1248** [explosion/spaCy#8812] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

I'm trying to train a textcat based on `en_core_web_sm`, without modifying any of the original weights.
Unfortunately there's not too many guides on how to train a textcat, and most of the existing guides are outdated and are based on spaCy v2 nor do any of them use config.cfg and `spacy train` which is the recommended method by the docs. 

Thus, I'm not sure if this is the right approach, but I'm certainly facing quite a few errors with unclear instructions along the way. Currently I'm using the following config.cfg:

```
[paths]
train = "data/pers_puns.train.spacy"
# train = "data/pers_puns.train.json"
dev = "data/pers_puns.dev.spacy"
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec", "tagger", "parser", "ner", "textcat"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[nlp.tokenizer]
@tokenizers = "spacy.Tokenizer.v1"

[components]

[components.tok2vec]
source = "en_core_web_sm"

[components.tagger]
source = "en_core_web_sm"

[components.parser]
source = "en_core_web_sm"

[components.ner]
source = "en_core_web_sm"

[components.textcat]
factory = "textcat"

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = ["tok2vec", "tagger", "parser", "ner"]
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
# [initialize.components.textcat.labels]
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]

[pretraining]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
tokenizer = {}
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.components.textcat]
labels = ["pers", "puns"]
```

And running `python -m spacy debug data config.cfg` gives me the error:

`TypeError: [E930] Received invalid get_examples callback in `Tok2Vec.initialize`. Expected function that returns an iterable of Example objects but got: []`

According to the docs having `init_tok2vec = null` should be allowed, since `init_tok2vec` is for "external resources to assist with transfer learning", which I am assuming is optional, but it looks like it should be non-empty? Also, in my dataset, I only tagged the categories of the document, leaving all information related to other components empty, since I'm only training the textcat, is this allowed?

## Your Environment

- **spaCy version:** 3.0.5
- **Platform:** Linux-5.8.0-55-generic-x86_64-with-glibc2.29
- **Python version:** 3.8.10
- **Pipelines:** en_core_web_md (3.0.0), en_core_web_sm (3.0.0)

**Comment by polm** [comment]
Sorry you're having trouble with this. 

I would recommend you first go through the official docs and use all the default settings to train a model, and then adjust it to keep what you want from the small model. In particular look at the [tutorial projects](https://github.com/explosion/projects/tree/v3/tutorials), mainly the goemotions one, to see an end-to-end working example of a textcat project.

Also since it sounds like this is a new project, I would recommend you upgrade to the latest spaCy, including upgrading your models. If there is a reason you don't want to do that please let me know so I can better understand your use case.

> According to the docs having init_tok2vec = null should be allowed, since init_tok2vec is for "external resources to assist with transfer learning", which I am assuming is optional, but it looks like it should be non-empty?

`init_tok2vec` can be null (and is in the sample projects). I think the issue here is that you're loading the tok2vec component from the small model, but since you're not using `init_tok2vec` something is weird with the initialization. 

> I'm trying to train a textcat based on en_core_web_sm, without modifying any of the original weights.

This is not possible in a straightforward way. spaCy models are normally trained with a tok2vec layer that is also updated, not just used for feature extraction. 

What you can do is train the textcat with its own tok2vec, and then load both of those into the small English pipeline as a component afterwards. Read the section in the docs on [Shared embedding layers](https://spacy.io/usage/embeddings-transformers#embedding-layers) for reference on how this works. 

Components should generally not be in your training config unless you are training them or using their annotations during training (like with a sentence tokenizer or something).

**Reply by kevinlu1248** [reply]
Thanks for the quick response. Sorry for the delay, I got caught up with a few things. Just to clarify, do you mean I should try training with everything, then cut down as necessary? I'm wondering this because I want to take advantage of transfer learning since my data set is small (300 examples per category).

**Reply by kevinlu1248** [reply]
> init_tok2vec can be null (and is in the sample projects). I think the issue here is that you're loading the tok2vec component from the small model, but since you're not using init_tok2vec something is weird with the initialization.

I'm using the same config from the samples (https://github.com/explosion/projects/tree/v3/tutorials/textcat_goemotions) but with different paths and I'm getting the same error when running `spacy debug data config.cfg`. Also forgot to mention but I am using spacy 3.0.5.

**Comment by svlandeg** [comment, marked as answer]
> TypeError: [E930] Received invalid get_examples callback in Tok2Vec.initialize. Expected function that returns an iterable of Example objects but got: []

The `get_examples` callback is a function that is passed through all components to initialize them with sample training data. It looks like the error you're getting, is telling you that no sample data (`[]`) could be parsed from the train/dev files you provided. This isn't really related to the `init_tok2vec` config setting, which is used for pretrained tok2vec weights, but I can definitely see how the naming is confusing.

Anyway - could you double check whether the `.spacy` files you specified are definitely accessible from that path (`"data/pers_puns.train.spacy"`), and that they contain valid data? You can read them in directly with `DocBin` to double check their contents, e.g.:

```
doc_bin = DocBin().from_disk("data/pers_puns.train.spacy")
vocab = English().vocab
docs = list(doc_bin.get_docs(vocab))
for doc in docs:
    print(doc.text, doc.cats)
``` 

(maybe the path needs a `./` or something?)

**Reply by kevinlu1248** [reply]
I tried the above and got no output. I think I found the issue, I'm getting 
```
✔ Generated output file (0 documents): data/pers_puns.dev.spacy
```
When converting my json file into spacy file. My json file is non-empty but in the "paragraphs" key for every element I have an empty list for the "sentences" key. This is because I'm only training the textcat and not the tokenizer and parser which the "sentences" key seems to be associated with, but I believe this assumption is wrong?

Also, I was wondering if we could change the `spacy convert` command line to yield an error on generating 0 documents to avoid future confusion?

**Reply by kevinlu1248** [reply]
Okay I found a bug in my data format which I fixed (not getting 0 documents anymore), and now running the code you sent above yields a list of the labels but not the raw texts. Not sure why the text isn't showing up since my "raw" fields are clearly non-empty. Further, I'm still getting the same error with directly training my model.

**Reply by polm** [reply]
If the text isn't showing up with the sample code from Sofie above that's a problem you'll need to fix. Can you give a sample of your JSON data, or maybe try skipping JSON and creating a DocBin directly? The preprocess scripts in the [example projects](https://github.com/explosion/projects/tree/v3/tutorials) should be a good reference.

**Reply by kevinlu1248** [reply]
I manually made the DocBin's and now they work! (based on https://github.com/explosion/projects/blob/v3/tutorials/textcat_goemotions/scripts/convert_corpus.py)

Thanks for all the help! Now onto the hyperperameter search.

**Reply by svlandeg** [reply]
Happy to hear you got it working!
